Feature/detailed review attachments view

miggol opened the pull request:
Reviewers still needed a more detailed attachments view, for example to display attachment names and comments. This PR should fill that gap.

Additionally I continued working on the comments from #736, which I will further address here.

I have *tried* to reduce the number of unique classes, however I have failed :-/

#### Items vs. Slots

As these classes look rather similar and do similar things, it might make sense to just use give slots the right methods and use slots everywhere. The main problem with this is that an item in the documents list does not always  represent an attachment. Sometimes it might be the PDF application, METC decision, or what have you. Those files do not live in a slot.

This is why the item class needs to exist as a wrapper around whatever thing is being put in the list.

Slots and items also have different purposes. Slots really are a stepper object. They represent a place to put an attachment, even if it's empty. And the DocItem (and the AttachmentItem wrapper) should represent any kind of document in a list.

#### Vision

So then when it comes to vision, I guess that I like having small wrapper classes because they make our code less interdependent. This makes it easier to change things down the road, and will hopefully also avoid us having to fix a million things at once in the future.

For example, by having made these kinds of container classes like `DocItem` when developing the previous documents_list, it was very easy to reuse code for the newer attachments list. The following class does seem a bit frivolous:

```
class AttachmentItem(DocItem):
    """
    A wrapper for AttachmentsList items that works with Attachment slots.
    """

    @property
    def attachment(
        self,
    ):
        return self.slot.attachment.get_correct_submodel()

    def get_link_url(
        self,
    ):
        return self.slot.attachment.get_download_url(
            self.slot.get_proposal(),
        )

    def get_filename(
        self,
    ):
        return self.attachment.upload.original_filename
```

But I really like that with just a few glue methods we can reuse existing code. Slots and items serve only slightly different purposes, but I still prefer to have them each be good at their own jobs and then write a bit of glue inbetween.

#### Other strong similarities / potential merges ion the future

There are strong similarities in how we fetch and sort attachments per attached object in the attachments list and in `ProposalAttachmentsList.get_context_data()`. That's kind of lame but I'm not yet convinced it's worth mashing these together.

I do think it's kind of attractive to mangle `DocList` into the stepper. It would be nice to have `stepper.slots.per_item` and `stepper.slots.as_containers` functions which we could use in many places. However, because we would still need to somehow shoehorn things like the METC decision into the review sidebar, I have left this as is for now. It's just not worth the effort as things are still developing. But it might be worth doing in the future.

#### Editing attachments as secretary

This is more of an emergency feature, hopefully it will be far less necessary now that the attachments system is improved. Because it should only be used as an exception, I'm fine with it being the same as the user's edit attachments view.

I also believe that it is good to only have one view to maintain, and it guarantees that there's no mismatch between the slots that the secretary sees and the slots that user sees.

It would be an improvement to remove the stepper from this page. But I just don't really see it as worth the effort right now.

#### Conclusion

* I have moved some methods such as `comparable()` and `is_new()` from the Item to the Slot class, but in general I have kept the number of classes the same.
* Wherever I have failed to reduce the number of classes, I have rearranged them in more logical places which will make things more maintainable going forward. For example in their own `reviews/utils/attachment_utils.py`
* Of course there's also a new detailed attachments view, check it out!

EdoStorm96 reviewed (changes requested):
Nice job. I think it looks good and works nicely. I have some changes I would like to see, but nothing major. See my comments.

Other general remarks:

- I guess the DMP page needs some work. We should probably remove the file field from this page? And ideally move it before the attachments page. But maybe this should be another issue/PR.

- So I would really prefer if the secretary attachments page did not include the stepper and navigation buttons. Can´t we just subclass the attachments view and add a variable to the context, like is_review_stage, and based on this, overwrite the sidebar block (maybe include the review_details_sidebar? Though maybe not.) and override the page_buttons block with a button redirecting to the review_detail?

miggol commented on proposals/views/attachment_views.py:
I think it's still good to show "no documents found" so that there are no trajectories missing from the list entirely if there are no documents there.

But if there are none at all, the legacy documents list shouldn't show at all either. So I've added an if statement to the review sidebar to make that section hide itself, like the proposal attachments view already did.

EdoStorm96 replied:
Sorry, I should have been clearer. Although I appreciate removing it from the review detail sidebar, I would also suggest removing it from the normal attachments page. Now we can have this:
![image](https://github.com/user-attachments/assets/211727d6-20ee-41ce-954f-0392ee899956)
In this case, I think the whole warning is just confusing if there are no legacy attachments, and I would prefer to get rid of it, when there are 0 legacy attachments. Because you won't be able to add them anyway.

miggol replied:
Damn I failed pretty badly at testing this. There will still be a list of empty containers so it is still truthy and will render. Thank you for being thorough!

miggol commented on attachments/models.py:
Done!

miggol commented on proposals/views/attachment_views.py:
```diff
@@ -2,8 +2,12 @@
 from proposals.mixins import ProposalContextMixin
 from proposals.models import Proposal
+from proposals.utils.stepper import Stepper
+from reviews.models import Review
```
Done

miggol commented on reviews/templates/reviews/review_attachments.html:
I've added a button at the top and bottom of the page. It might be a bit much, but I couldn't really figure out how to style it better.

miggol commented:
Alright, thanks for the review. I've addressed all of your comments except for removing the DMP page, for which I've created another issue (#756). It may seem trivial but I would rather get this attachments page on the test server sooner.

At first I thought removing the stepper would be super annoying, because it would involve seven extra URL paths which we would have to remember to maintain every time we change any of the attachment pages. But after your second request I went looking for an alternative and came up with a mixin solution.

It's super hacky in that it just checks if the current user is secretary and not an applicant/supervisor then overwrites the stepper. But I consider it fit for purpose.

EdoStorm96 commented on proposals/templates/proposals/attachments.html:
```diff
@@ -55,3 +55,7 @@ <h6>{{ container.header }}</h6>
+
+{% if secretary_return_link %}
+    {% block form-buttons %}{% endblock %}
+{% endif %}
```
Somehow, this is not working for me. Even though I can plainly see that this variable is not in the context, it now won't show the normal form-buttons when I'm in the form. I am having succes with putthing this in fetc_form_base.html
```
                            {% block form-buttons %}
                                {% if not secretary_return_link %}
                                    {% include "base/form_buttons.html" %}
                                {% endif %}
                            {% endblock %}
```
and then removing this stuff from attachments.html

miggol replied:
Aargh my bad, templates don't work like that. Because the if statement (as I wrote it) is outside of any block, it never gets evaluated. However, Django does still consider the form-buttons to be overwritten, probably because it uses some other kind of mechanism to determine that. So they were always overridden.

Your solution works! I adjusted it to use `{{ block.super }}` though, so it's more dependable.

EdoStorm96 reviewed (approved):
I have two more issues ... Removing the buttons from attachments.html does not seem to work as expected. I have suggested a fix. And see my comment about the other legacy documents :)

If you fix this, I approve. Nice job!
